basculehttp/bearerTokenFactory_test.go

```diff
@@ -17,154 +17,226 @@
+			parseCalled: true,
+			parseErr:    parseFailErr,
+			expectedErr: parseFailErr,
+		},
+		{
+			description:   "Resolve Key Error",
+			value:         "abcd",
+			parseCalled:   true,
+			resolveCalled: true,
+			resolveErr:    resolveFailErr,
+			expectedErr:   resolveFailErr,
+		},
+		{
+			description:   "Invalid Token Error",
+			value:         "abcd",
+			parseCalled:   true,
+			resolveCalled: true,
+			validToken:    false,
+			expectedErr:   ErrInvalidToken,
+		},
+		{
+			description:   "Convert to Claims Error",
+			value:         "abcd",
+			parseCalled:   true,
+			resolveCalled: true,
+			validToken:    true,
+			expectedErr:   ErrUnexpectedClaims,
+		},
+		{
+			description:   "Get Principal Error",
+			value:         "abcd",
+			parseCalled:   true,
+			resolveCalled: true,
+			validToken:    true,
+			claims:        &bascule.ClaimsWithLeeway{},
+			expectedErr:   ErrInvalidPrincipal,
+		},
+		{
+			description:   "Non-string Principal Error",
+			value:         "abcd",
+			parseCalled:   true,
+			resolveCalled: true,
+			validToken:    true,
+			claims: &bascule.ClaimsWithLeeway{
+				MapClaims: jwt.MapClaims{jwtPrincipalKey: 55.0},
+			},
+			expectedErr: ErrInvalidPrincipal,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.description, func(t *testing.T) {
+			assert := assert.New(t)
+			r := new(key.MockResolver)
+			p := new(mockParser)
+			pair := new(key.MockPair)
+			if tc.parseCalled {
+				token := jwt.NewWithClaims(jwt.SigningMethodHS256, tc.claims)
+				token.Valid = tc.validToken
+				p.On("ParseJWT", mock.Anything, mock.Anything, mock.Anything).Return(token, tc.parseErr).Once()
+			}
+			if tc.resolveCalled {
+				r.On("ResolveKey", mock.Anything, mock.Anything).Return(pair, tc.resolveErr).Once()
+				if tc.resolveErr == nil {
+					pair.On("Public").Return(nil).Once()
+				}
+			}
+			btf := BearerTokenFactory{
+				DefaultKeyID: "default key id",
+				Resolver:     r,
+				Parser:       p,
+			}
+			req := httptest.NewRequest("get", "/", nil)
+			token, err := btf.ParseAndValidate(context.Background(), req, "", tc.value)
+			assert.Equal(tc.expectedToken, token)
+			if tc.expectedErr == nil || err == nil {
+				assert.Equal(tc.expectedErr, err)
+			} else {
+				assert.Contains(err.Error(), tc.expectedErr.Error())
+			}
+		})
+	}
+}
+
+func TestProvideBearerTokenFactory(t *testing.T) {
+	type In struct {
+		fx.In
+		Options []COption `group:"bascule_constructor_options"`
+	}
+
+	const yaml = `
+good:
+  key:
+    factory:
+      uri: "http://test:1111/keys/{keyId}"
+    purpose: 0
+    updateInterval: 604800000000000
+`
+	v := viper.New()
+	v.SetConfigType("yaml")
+	require.NoError(t, v.ReadConfig(strings.NewReader(yaml)))
+
+	tests := []struct {
+		description    string
+		key            string
+		optional       bool
+		optionExpected bool
+		expectedErr    error
+	}{
+		{
+			description:    "Success",
+			key:            "good",
+			optional:       false,
+			optionExpected: true,
+		},
+		// {
```

Comment:
did you mean to remove this case?

Reply:
Oh yeah, that case is currently impossible to hit so it can't be tested.  I'll remove it.